### PR TITLE
feat: remove Schema functor variations that depend on external modules

### LINKED ISSUE #36: Suggestion: remove 'variations'
#### Motivation

Reduce the dependency footprint. Right now, due to https://github.com/RationalJS/future/issues/53, I cannot update my bs-platform version to 8.

Since these variations are very simple and short, they could be documented instead of being a part of the source code. There's also the option of making each variation a separate npm module, but it might be overkill.

I'll be happy to submit a PR if you agree with this suggestion.

### PR DESCRIPTION
**Pull request by leeor**
BREAKING CHANGE

This library no longer provides instantiations of the Schema functor
based on `reason-promise` and `reason-future`.

Fixes #36.

The approach of providing an interface file for `Graphql_Schema.re` using another `.re` file that provides the type of the instantiated functor was taken from the OCaml Graphql server this implementation was originally based on.
However, I did not trim down the types and bindings `Schema` exposes. It's probably a good idea to do so, and this is a good time to do it, but I think you should have a better understanding of what needs to be exposed and what does not. My subjective use case should not be the benchmark.

Since this is a breaking change, it might be a good chance to turn on `namespace` in `bsconfig.json`.

### CODE REVIEW

File: reason-graphql/src/Graphql_Schema.re [hunk 1]
```diff
@@ -92,6 +81,17 @@ module Make = (Io: IO) => {
...
+    include Belt.Map.String;
+    exception MissingKey(string);
+  };
+
+  module StringSet = Set.Make(String);
+
+  type variableList = list((string, Ast.constValue));
+  type variableMap = StringMap.t(Ast.constValue);
+  type fragmentMap = StringMap.t(Ast.fragmentDefinition);
+
```
Comment (sikanhe) [inline reason-graphql/src/Graphql_Schema.re:94]: I don't get this change

File: reason-graphql/src/Graphql_Interface.re [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+  let return: 'a => t('a);
+  let bind: (t('a), 'a => t('b)) => t('b);
+  let map: (t('a), 'a => 'b) => t('b);
+};
+
+module type Schema = {
+  module Io: IO;
+
+  module StringMap: {
+    type key = string;
```
Comment (sikanhe) [inline reason-graphql/src/Graphql_Interface.re:12]: Why is this needed? and Why do we need to expose StringMap type?

Comment (leeor) [reply reason-graphql/src/Graphql_Interface.re:12]: This is needed so we have the type of the module returned from the `Schema` functor.

We need that type so that consumers of the library are able to refer to it directly, and not through their own implementation that instantiates the functor. This definition is agnostic to the async library they choose to use.

As for `StringMap`, the `Belt.Map` module does not expose an `S` module type for the module returned from the map functor, so we don't have a choice but to "type" it all. There are several type definitions that use `StringMap`: `variableMap`, `fragmentMap`, and a few functions that deal with the collection and validation of fragments. If none of these needs to be exposed then we can remove the verbose definition of the `StringMap` module.

File: reason-graphql/src/Graphql_Schema.re (hunk 1 above)
Comment (leeor) [reply reason-graphql/src/Graphql_Schema.re:94]: The definition of the `Schema` module type in `Graphql_Interface.re` required that `StringMap` and the three other types be moved "into" the `Schema` module. It was the easiest way to resolve the typing issue.
It then necessitated the inclusion of these in the implementation file too.

File: reason-graphql/src/Graphql_Interface.re (hunk 2 above)
Comment (sikanhe) [reply reason-graphql/src/Graphql_Interface.re:12]: I think these are only used internally for resolving executions. 

Are we sure we cannot do something like `include (module type of FunctorReturnType) with type key = string` ?

**Comment by sikanhe** [comment]
@leeor so are you saying that we cannot remove the variations without doing this refactor?

File: reason-graphql/src/Graphql_Interface.re (hunk 2 above)
Comment (leeor) [reply reason-graphql/src/Graphql_Interface.re:12]: We can't `include type of FunctorReturnType` because there is not module type for the functor return type in `Belt.Map`. I am not familiar with a way to deduce a functor's returned module type.

If all the uses of `StringMap` are internal, I'll happily remove its definition and anything that uses it. However, I don't think I can make this decision on the basis of my own single use case.

**Comment by leeor** [comment]
It's completely feasible to separate into two PRs, one for the schema interface, one for actually removing the variations.

I'll update over the next few days.

**Comment by leeor** [comment]
@sikanhe, this can be applied after merging #40 which is a separate PR for adding the schema interface.

Per your request, I've removed the StringMap module and any other types/bindings that depend on it.

**Comment by sikanhe** [comment]
Done in https://github.com/sikanhe/reason-graphql/commit/03335d8c911aaec58ebd33c54d38df032ee1fcc8

**Comment by leeor** [comment]
This PR was already rebased on top of the schema interface change, so no need to rebase.

**Comment by leeor** [comment]
@sikanhe any reason for closing this PR without merging it?
